<a href="https://colab.research.google.com/github/ImamSyabana/ML-terapan/blob/main/FinalSubmisi_MLterapan_updated.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Proyek akhir Machine Learning Terapan : Muhammad Imam Ariq Sya'bana

# Data Understanding

### Data Wrangling

In [1]:
from google.colab import drive
drive.mount('/content/drive')


Mounted at /content/drive


Menyaipkan semua library yang dibutuhkan

In [2]:
#masukan librari
import numpy as np
import pandas as pd
import scipy
import matplotlib.pyplot as plt
import seaborn

### Gathering Data

Al-Qur'an terjemahan datasets : https://www.kaggle.com/datasets/zusmani/the-holy-quran/data?select=Indonesian.csv

Datasets Link from Gdrive : https://drive.google.com/file/d/1SODjkj2OYZkMvtq9TMXJ-OK8HoqKvVxq/view?usp=sharing


In [3]:
dataset = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Semester 7/QURAN HADIS MINING/dataset indeks quran/DataSet-Terjemah Al Quran - Copy.csv')


memvisualisasikan dataframe

In [4]:
dataset

Surat  Ayat                                           terjemah
0         1     1  Dengan menyebut nama Allah Yang Maha Pemurah l...
1         1     2          Segala puji bagi Allah Tuhan semesta alam
2         1     3                   Maha Pemurah lagi Maha Penyayang
3         1     4                    Yang menguasai hari pembalasan.
4         1     5  Hanya kepada Engkaulah kami menyembah dan hany...
...     ...   ...                                                ...
6229    114     2                                      Raja manusia.
6230    114     3                                  Sembahan manusia.
6231    114     4  dari kejahatan (bisikan) setan yang biasa bers...
6232    114     5  yang membisikkan (kejahatan) ke dalam dada man...
6233    114     6                   dari (golongan) jin dan manusia.

[6234 rows x 3 columns]

### Assessing Data

mendeskripsikan metadata dari dataset yang digunakan

In [5]:
print("Jumlah Surat : ", len(dataset['Surat'].unique()))


Jumlah Surat :  114


Memahami atribut-atribut yang ada di dalam dataset tersebut dilakukan dengan beberapa langkah untuk memahami isi dan tipe atribut tersebut.

In [6]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6234 entries, 0 to 6233
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Surat     6234 non-null   int64 
 1   Ayat      6234 non-null   int64 
 2   terjemah  6234 non-null   object
dtypes: int64(2), object(1)
memory usage: 146.2+ KB


Menghitung data yang unik pada fitur Surat Al-Qur'an yang ada pada dataset ini untuk mengecek semua surat lengkap .

In [7]:
print("Jumlah paper : ", len(dataset.Surat.unique()))
print("\nContoh Salah satu Ayat dalam Al-Qur'an di dataset : ", dataset['terjemah'][0])

Jumlah paper :  114

Contoh Salah satu Ayat dalam Al-Qur'an di dataset :  Dengan menyebut nama Allah Yang Maha Pemurah lagi Maha Penyayang


jumlah ayat al-qur'an terjemahan di dalam dataset tersebut berjumlah 6234 dengan ayat yang unik sebanyak 6160.

In [8]:
print("Jumlah Surat terjemahan : ", len(dataset.terjemah))
print("Jumlah Surat terjemahan Uniq : ", len(dataset.terjemah.unique()))

Jumlah Surat terjemahan :  6234
Jumlah Surat terjemahan Uniq :  6160


memeriksa apakah ada data yang memiliki missing value

In [9]:
# Cek missing value dengan fungsi isnull()
dataset.isnull().sum()

Surat       0
Ayat        0
terjemah    0
dtype: int64

karena tidak terdapat missing value, sehingga tidak perlu dilanjutkan ke tahap penanganan missing value seperti drop missing value, imputation, dan interpolation

In [10]:
# # Membersihkan missing value dengan fungsi dropna()
# dataset = dataset.dropna()
# dataset

mengecek apakah terdapat dua atau lebih data yang memiliki nilai yang sama pada semua atributnya (mengecek duplikat)

In [11]:
print("Jumlah duplikasi: ", dataset.duplicated().sum())

Jumlah duplikasi:  0


karena tidak terdapat nilai duplicate, sehingga tidak perlu dilanjutkan ke tahap penanganan duplicate value seperti drop_duplicates

In [12]:
# # Membuang data duplikat pada variabel preparation
# dataset = dataset.drop_duplicates()
# dataset

In [13]:
dataset.sample(5)


Surat  Ayat                                           terjemah
2111     17    85  Dan mereka bertanya kepadamu tentang roh. Kata...
4186     40    56  Sesungguhnya orang-orang yang memperdebatkan t...
1521     11    51  Hai kaumku aku tidak meminta upah kepadamu bag...
5412     70    40  Maka Aku bersumpah dengan Tuhan Yang Mengatur ...
441       3   150  Tetapi (ikutilah Allah) Allahlah Pelindungmu d...

# Data Preprocessing

melakukan menganalisis data dan memanipulasi data

Memetakan semua atribut yang ada di dataset ke dalam variabelnya masing-masing

In [14]:
variabelSurat = dataset.iloc[:,0].values
variabelAyat = dataset.iloc[:, 1].values
variabelTerjemah = dataset.iloc[:, 2].values


In [15]:
type(variabelSurat)

numpy.ndarray

In [16]:
variabelAyat[:5]

array([1, 2, 3, 4, 5])

In [17]:
variabelTerjemah[:5]

array(['Dengan menyebut nama Allah Yang Maha Pemurah lagi Maha Penyayang',
       'Segala puji bagi Allah Tuhan semesta alam',
       'Maha Pemurah lagi Maha Penyayang',
       'Yang menguasai hari pembalasan.',
       'Hanya kepada Engkaulah kami menyembah dan hanya kepada Engkaulah kami mohon pertolongan'],
      dtype=object)

Mengubah tipe data variabel masing-masing atribut dari numpy.ndarray menjadi list

In [18]:
Surat_list = [] # rubah dari numpy array ke list
for i in range(len(variabelSurat)):
  Surat_list.append(variabelSurat[i])

Ayat_list = [] # rubah dari numpy array ke list
for i in range(len(variabelAyat)):
  Ayat_list.append(variabelAyat[i])

Terjemahan_list = [] # rubah dari numpy array ke list
for i in range(len(variabelTerjemah)):
  Terjemahan_list.append(variabelTerjemah[i])


In [19]:
Surat_list[-5:], Ayat_list[-5:], Terjemahan_list[-5:]

([114, 114, 114, 114, 114],
 [2, 3, 4, 5, 6],
 ['Raja manusia.',
  'Sembahan manusia.',
  'dari kejahatan (bisikan) setan yang biasa bersembunyi',
  'yang membisikkan (kejahatan) ke dalam dada manusia.',
  'dari (golongan) jin dan manusia.'])

In [20]:
print(type(Terjemahan_list))

<class 'list'>


# Modelling

Metode yang digunakan untuk tahap pembuatan model machine learning untuk projek sistem temu kembali rekomendasi ayat Al-Qur'an relevan adalah cosine similarity.

## mamasukkan querry yang ingin dicari


In [21]:
query = input("Masukkan querry dari judul paper yang ingin dicari : ")


Masukkan querry dari judul paper yang ingin dicari : Riba


## Melakukan Cosine similarity

In [22]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity


# preprocessing dan vektorisasi documents
vectorizer = TfidfVectorizer()
vectorized_docs = vectorizer.fit_transform(Terjemahan_list)


# preprocessing dan vektorisasi querry
query_vector = vectorizer.transform([query])

# Menghitung cosine similarity antara kueri dengan masing-masing ayat Al-Qur'an
# Calculate cosine similarity between the query and each document
similarity_scores = cosine_similarity(query_vector, vectorized_docs)

# Menampilkan similarity scores untuk setiap dokumen

for doc_idx, score in enumerate(similarity_scores[0]):
    surat = Surat_list[doc_idx]
    ayat  = Ayat_list[doc_idx]
    terjemah = Terjemahan_list[doc_idx]

    print(f"Similarity dengan Al-Qur'an pada \nSurat ke-{surat}, \nAyat ke-{ayat}, \nyaitu sebesar: {score:.5f} \nIsi Terjemahan:\n{terjemah}\n\n")

Streaming output truncated to the last 5000 lines.
Similarity dengan Al-Qur'an pada 
Surat ke-76, 
Ayat ke-21, 
yaitu sebesar: 0.00000 
Isi Terjemahan:
Mereka memakai pakaian sutera halus yang hijau dan sutera tebal dan dipakaikan kepada mereka gelang terbuat dari perak dan Tuhan memberikan kepada mereka minuman yang bersih.


Similarity dengan Al-Qur'an pada 
Surat ke-76, 
Ayat ke-22, 
yaitu sebesar: 0.00000 
Isi Terjemahan:
Sesungguhnya ini adalah balasan untukmu dan usahamu adalah disyukuri (diberi balasan).


Similarity dengan Al-Qur'an pada 
Surat ke-76, 
Ayat ke-23, 
yaitu sebesar: 0.00000 
Isi Terjemahan:
Sesungguhnya Kami telah menurunkan Al Qur'an kepadamu (hai Muhammad) dengan berangsur-angsur.


Similarity dengan Al-Qur'an pada 
Surat ke-76, 
Ayat ke-24, 
yaitu sebesar: 0.00000 
Isi Terjemahan:
Maka bersabarlah kamu untuk (melaksanakan) ketetapan Tuhanmu dan janganlah kamu ikuti orang yang berdosa dan orang yang kafir di antara mereka.


Similarity dengan Al-Qur'an pada 
Sur

In [23]:
query

'Riba'

In [24]:
similarity_scores[0]

array([0., 0., 0., ..., 0., 0., 0.])

## menyimpulkan kategori dari querry yang dimasukkan

In [33]:
# dictionary variabel yang menyimpan list
dict = {'Urutan Surat' : Surat_list, 'Urutan Ayat': Ayat_list, 'Terjemahan Indonesia' : Terjemahan_list, 'nilai kesamaan': similarity_scores[0]}

df_Quran = pd.DataFrame(dict)
df_Quran = df_Quran[df_Quran['nilai kesamaan'] != 0]
Quran_sorted = df_Quran.sort_values(by=['nilai kesamaan'], ascending=False)

Quran_sorted[:]

Urutan Surat  Urutan Ayat  \
280              2          275   
284              2          279   
3445            30           39   
421              3          130   
283              2          278   
281              2          276   
651              4          161   

                                   Terjemahan Indonesia  nilai kesamaan  
280   Orang-orang yang makan (mengambil) riba tidak ...        0.570177  
284   Maka jika kamu tidak mengerjakan (meninggalkan...        0.467033  
3445  Dan sesuatu riba (tambahan) yang kamu berikan ...        0.425930  
421   Hai orang-orang yang beriman janganlah kamu me...        0.361682  
283   Hai orang-orang yang beriman bertakwalah kepad...        0.351942  
281   Allah memusnahkan riba dan menyuburkan sedekah...        0.340732  
651   dan disebabkan mereka memakan riba padahal ses...        0.280317

In [31]:
type(Quran_sorted)

pandas.core.frame.DataFrame

# menampilkan semua hasil rekomendasi akhir dari sistem temu kembali

In [27]:
(Quran_sorted)

Urutan Surat  Urutan Ayat  \
280              2          275   
284              2          279   
3445            30           39   
421              3          130   
283              2          278   
281              2          276   
651              4          161   

                                   Terjemahan Indonesia  nilai kesamaan  
280   Orang-orang yang makan (mengambil) riba tidak ...        0.570177  
284   Maka jika kamu tidak mengerjakan (meninggalkan...        0.467033  
3445  Dan sesuatu riba (tambahan) yang kamu berikan ...        0.425930  
421   Hai orang-orang yang beriman janganlah kamu me...        0.361682  
283   Hai orang-orang yang beriman bertakwalah kepad...        0.351942  
281   Allah memusnahkan riba dan menyuburkan sedekah...        0.340732  
651   dan disebabkan mereka memakan riba padahal ses...        0.280317

In [28]:

# Export the DataFrame to a CSV file
csv_filename = 'Rekomendasi_ayat_Quran.csv'
Quran_sorted.to_csv(csv_filename, index=False, encoding='utf-8', header=True)